### ISSUE TITLE
Missing sample for image multiclass multilabel classification

### ISSUE THREAD

**Issue by codeclash** [dotnet/machinelearning-samples#585]
I've seen a couple of open/closed issues (both here in this samples repo and in [dotnet/machinelearning](https://github.com/dotnet/machinelearning) related to multiclass multilabel classification, but it's not clear to me how to actually implement it:

For example, I want to modify [TensorFlow image classifier with transfer learning and ML.NET](https://github.com/dotnet/samples/tree/master/machine-learning/tutorials/TransferLearningTF) in order to apply multilabelling (a single image can be assigned multiple labels instead of just one).

It is not clear - from any docs (api docs, roadmaps, release notes) or samples -  if this is actually feasible atm. with **ML.Net 1.2**.

Could you clarify if this can be already achieved with the current version, and if so, how, or if not, when to expect addition of those apis?

**Comment by CESARDELATORRE** [comment]
@codeclash - I think you can do what you are looking for. 

**Case A: Multiclass and multilabel by just scoring a TF model (such as INCEPTION) with its original labels/classes**

The base TensorFlow models such as INCEPTION already support multilabel since you can see in the TF scoring sample that you get a list of all the labels with a different % probability per each. Debug/F5 this code and you'll see the list of probabilities: 

https://github.com/dotnet/machinelearning-samples/blob/master/samples/csharp/getting-started/DeepLearning_ImageClassification_TensorFlow/ImageClassification/ModelScorer/TFModelScorer.cs#L97  

**Case B: Transfer learning on top of a TF model (such as INCEPTION) by training with your own labels/classes with ML.NET and a multi-class classification trainer:**

Going forward, if you want to do transfer learning and train with ML.NET (such as using a multi-class classification trainer) on top of the TensorFlow model, you can also do it and have multilabel, as well.

Basically, it'd be a similar sample than the TF-Estimator sample where you train a multi-class classification trainer on top of the TF model-estimator:

https://github.com/dotnet/machinelearning-samples/tree/master/samples/csharp/getting-started/DeepLearning_TensorFlowEstimator

But in addition to that, then you can take similar code for **multilabel** score/percentage list (meaning that a single image can be related to multiple labels/classes and you can have certain % of probability per each) by accessing the schema info, such as in the 'GitHub issues classification' sample, here:

In that sample it can also use an LbfgsMaximumEntropy trainer for multi-class classification: 

`.Append(mlContext.MulticlassClassification.Trainers.LbfgsMaximumEntropy(labelColumnName:LabelTokey, featureColumnName:"softmax2_pre_activation"))`

The code you are probably looking for is here:

https://github.com/dotnet/machinelearning-samples/blob/master/samples/csharp/end-to-end-apps/MulticlassClassification-GitHubLabeler/GitHubLabeler/GitHubLabelerConsoleApp/Labeler.cs#L63

```
            //Predict labels and scores for single hard-coded issue
            var prediction = _predEngine.Predict(singleIssue);
            _fullPredictions = GetBestThreePredictions(prediction);
            Console.WriteLine("==== Displaying prediction of Issue with Title = {0} and Description = {1} ====", singleIssue.Title, singleIssue.Description);
            Console.WriteLine("1st Label: " + _fullPredictions[0].PredictedLabel + " with score: " + _fullPredictions[0].Score);
            Console.WriteLine("2nd Label: " + _fullPredictions[1].PredictedLabel + " with score: " + _fullPredictions[1].Score);
            Console.WriteLine("3rd Label: " + _fullPredictions[2].PredictedLabel + " with score: " + _fullPredictions[2].Score);
```

The important code is within our custom `GetBestThreePredictions(prediction);` method where you use schema APIs to get the list of the classes with a % of probability per each. Then we're returning the best 3:

https://github.com/dotnet/machinelearning-samples/blob/master/samples/csharp/end-to-end-apps/MulticlassClassification-GitHubLabeler/GitHubLabeler/GitHubLabelerConsoleApp/Labeler.cs#L74

Please, confirm if any of those cases is what you are looking for. :)

**Comment by codeclash** [comment]
@CESARDELATORRE thanks so much for your input! I've already looked at the samples you mentioned, but  I don't think I'll by able to solve what I want purely with ML.Net, see [TensorflowCatalog Class](https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.tensorflowcatalog?view=ml-dotnet):

> 3. Advanced transfer learning/fine tuning scenarios (e.g. adding more layers into the network, changing the shape of inputs, freezing the layers which do not need to be updated during retraining process etc.) are currently not possible due to lack of support for network/graph manipulation inside the model using TensorFlow C-API.

_Also I'm not an ML expert, so I hope all this makes sense to you :-)_

After reading an [example using inception and doing multilabelling in pyhton](https://towardsdatascience.com/multi-label-image-classification-with-inception-net-cbb2ee538e30) I would have to dynamically modify the inception model so it can be trained on N number of classes per training run (I would like to be able to define N classes on a webui client, do manual labelling on the client, and then submit to the ML server, which sets up for that N classes, takes the multilabelled images, trains with them, and then additionally also predicts some M additional images and returns those to the client. 

The idea is kind of an **assisted human labelling tool** via WebInterface (I have this implemented already for multiclass with ML.Net and Blazor entirely, but I can only assign one class to an image instead of multiple labels):  

1. Via Webinterface, User defines a first bunch of classes he's sure about.
2. Webinterface fetches whole list of unlabelled images (urls) from backend, and presents a subset to the user as "unlabelled".
3. User starts (multi-)labelling yet "unlabelled" images. If he discovers that a new class (label) should be defined, he can just add it.
4. Once there are enough labelled images per class, user can hit "AutoClassify", which will send the labelled images (just imagename/labels) to the ML server, which will start a fresh run:
4.1 ML Server: prepare the model for N classes.
4.2 Train with the labelled images
4.3 Predict for a number of yet unlabelled images (say 100, configurable on the client), and return the results (predicted images including scores/labels/image + perf metrics)
5. The user can review the predictions, and if needed correct wrong predictions, by reassigning. 
6. Depending on the quality of the predictions, user can now either continue with 3.) directly with 4.) or go to 7.)
7. Once predictions are good and satisfying, user can hit "Save", and this will save & export the trained model (last run) so it can be used as a pretrained model somewhere else.

The whole purpose of this is to significantly speed up time spent on upfront labelling, since most time will be only spent in reviewing predictions (which should get better on each run) instead of actually manually assigning labels. Of course UX-wise this is only reasonable for a small number of classes.

Now, to enable this "dynamic" behaviour on the ML side, I need to be able to modify the inception model on the fly, eg. referring to [example using inception and doing multilabelling in pyhton](https://towardsdatascience.com/multi-label-image-classification-with-inception-net-cbb2ee538e30)  change the ground truth vectors and replace the softmax function with a sigmoid function because

> The softmax function squashes all values of a vector into a range of [0,1] summing together to 1. Which is exactly what we want in a single-label classification. But for our multi-label case, we would like our resulting class probabilities to be able to express that an image of a car belongs to class car with 90% probability and to class accident with 30% probability etc. 

As mentioned on top, ML.Net doesn't support this atm, so we'll to go to pyhton, and **I think this issue should be therefore closed**, right?

**Comment by CESARDELATORRE** [comment]
If you need to do transfer learning on multiple TensorFlow layers (not just add an additional ML.NET trainer on the ML.NET pipeline like our current sample) then you have the following options:

**- Using .NET:**
  - Use **TensorFlow.NET**. You have a transfer learning sample here: https://github.com/SciSharp/TensorFlow.NET/blob/master/test/TensorFlowNET.Examples/ImageProcessing/RetrainImageClassifier.cs

  - From the ML.NET team we're currently creating a higher level API on top of TensorFlow.NET so you'll be able to easily do transfer learning on one or multiple layers with a simpler **'ImageClassificationTrainer' API in ML.NET**. This is an initial PR with a 'specs doc' you can start taking a look, so far only to provide feedback, though:
https://github.com/dotnet/machinelearning/pull/4037

We expect to have an early preview version of that high level 'ImageClassificationTrainer' API in the next minor version: ML.NET 1.3, in August.

**- Using Python**
   - Use Keras
   - USing Tensorflow API from Python

**Comment by codeclash** [comment]
@CESARDELATORRE perfect, thanks you for your great support. 
ML:NET 1.3 sounds promising :-). In the meantime my colleague will implement it in python as usual, I'll see if I it makes sense to then port it to Tensorflow.NET once he's done, and will recheck in detail what  ML.NET 1.3 brings. It would be nice if the whole thing was dotnet end-to-end :-)

**Comment by luisquintanilla** [comment]
Closing this issue since a response was provided and there has been no additional activity

**Comment by Oceania2018** [comment]
@codeclash The TensorFlow.NET implementation moved to https://github.com/SciSharp/SciSharp-Stack-Examples/blob/master/src/TensorFlowNET.Examples/ImageProcessing/TransferLearningWithInceptionV3.cs.
